Lecture_12/Homework_12.py

```diff
@@ -0,0 +1,71 @@
+# 1. Використати файл як базу даних для збереження записів телефонної книги із попередніх завдань.
+
+# Ваша телефонна книга, що до цього містилася в dict, має зберігатися у вигляді тексту в JSON форматі.
+# При закритті програми і повторному відкритті всі попередні дані мають бути доступними.
+# Підказка: Ви можете використати бібліотеку json, яка допоможе із перетворенням dict в JSON string (при записі в файл)
+# і JSON string в dict (при читанні із файлу). Файл слід оновлювати після кожної успішної операції add або delete.
+
+# 2. Написати декоратор, який буде записувати в файл назву функції, яку він декорує, і писати час її виклику.
+
+# 3. В попередньо написаний кастомний Exception додати запис помилки і час її виникнення у файл.
+
+import json
+from datetime import datetime
+
+
+class MyCustomException(Exception):
+    with open('log.txt', 'a') as log_file:
+        print(f'\nFunction {Exception} is called at {datetime.now()}', file=log_file) # записуємо в log.txt виклик функції
+    pass
+
+def my_deco(func):
+    def wrap(*args, **kwargs):
+        with open('log.txt', 'a') as log_file:
+            print(f'\nFunction {open} is called at {datetime.now()}', file=log_file) # записуємо в log.txt виклик функції
+        func(*args, **kwargs)
+        return func(*args, **kwargs)
+    wrap
+
+@my_deco
+def my_func():
+    print('my_func')
+
+with open('json_dict.txt',) as phone_data: #Зчитємо дані з файла
```

Comment:
if data in the file has type `json` - the file name should use the extension` .json`
nit: good practice using comments: two spaces after code + # + space
nit: all time set in what `mode` you open file

```suggestion
with open('contacts.json', 'r') as phone_data:  # Зчитуємо дані з файла
```